engine/src/main/java/org/hibernate/search/v6poc/engine/impl/SearchMappingRepositoryBuilderImpl.java

```diff
@@ -142,34 +146,62 @@ public SearchMappingRepository getBuiltResult() {
 	private static class MapperContribution<C, M extends MappingImplementor> {
 
 		private final MapperFactory<C, M> mapperFactory;
+		private final Set<MappableTypeModel> typesLeftToProcess = new LinkedHashSet<>();
+		private final Set<MappableTypeModel> typesBeingProcessed = new LinkedHashSet<>();
+		private final Set<MappableTypeModel> freezedTypes = new HashSet<>();
 		private final Map<MappableTypeModel, TypeMappingContribution<C>> contributionByType = new HashMap<>();
 
 		MapperContribution(MapperFactory<C, M> mapperFactory) {
 			this.mapperFactory = mapperFactory;
 		}
 
 		public void update(MappableTypeModel typeModel, String indexName, C contributor) {
+			if ( freezedTypes.contains( typeModel ) ) {
+				throw new AssertionFailure(
+						"Attempt to add a mapping contribution for a type which has already been mapped."
+						+ " There is a bug in the mapper implementation."
+				);
+			}
+			typesLeftToProcess.add( typeModel );
 			contributionByType.computeIfAbsent( typeModel, TypeMappingContribution::new )
 					.update( indexName, contributor );
 		}
 
 		public Mapper<C, M> preBuild(BuildContext buildContext, ConfigurationPropertySource propertySource,
 				IndexManagerBuildingStateHolder indexManagerBuildingStateHolder) {
 			Mapper<C, M> mapper = mapperFactory.createMapper( buildContext, propertySource );
-			for ( MappableTypeModel typeModel : contributionByType.keySet() ) {
-				Optional<String> indexNameOptional = typeModel.getAscendingSuperTypes()
-						.map( contributionByType::get )
-						.filter( Objects::nonNull )
-						.map( TypeMappingContribution::getIndexName )
-						.filter( Objects::nonNull )
-						.findFirst();
-				if ( indexNameOptional.isPresent() ) {
-					String indexName = indexNameOptional.get();
-					mapper.addIndexed(
-							typeModel,
-							indexManagerBuildingStateHolder.startBuilding( indexName ),
-							this::getContributors
-					);
+
+			/*
+			 * We have to loop, and copy the set of types before starting processing,
+			 * because we allow mappers to add new contributions to new types
+			 * when we call org.hibernate.search.v6poc.entity.mapping.building.spi.Mapper.addIndexed.
+			 * This is necessary to allow annotation-based mapping,
+			 * which may discover new embedded types while mapping a type.
+			 *
+			 * We do not, however, allow new contributions to types already encountered in this method
+			 * (the "freezed" types).
+			 */
+			while ( !typesLeftToProcess.isEmpty() ) {
+				typesBeingProcessed.clear();
+				typesBeingProcessed.addAll( typesLeftToProcess );
+				typesLeftToProcess.clear();
+				freezedTypes.addAll( typesBeingProcessed );
```

Comment:
I don't think it has importance but aren't the types freezed after the execution of the following code.

I don't think it changes anything but it would seem more logical to me.

Reply:
They are actually freezed before, because as soon as we started processing them, we don't want any more contribution to these types. The "freezing" here is about rejecting new contributions, not about marking the contributions as processed.

Reply:
I will make it more explicit in the error message.